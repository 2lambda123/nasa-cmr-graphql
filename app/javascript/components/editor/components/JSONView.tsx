import { cloneDeep } from 'lodash'
import { observer } from 'mobx-react'
import React from 'react'
import {
  Accordion, Card
} from 'react-bootstrap'
import './JSONView.css'
import JSONPretty from 'react-json-pretty'
import MetadataEditor from '../MetadataEditor'
import { removeEmpty } from '../utils/json_utils'

type JSONViewProps = {
  editor: MetadataEditor;
}
type JSONViewState = {
  icon: string
}

class JSONView extends React.Component<JSONViewProps, JSONViewState> {
  constructor(props: JSONViewProps) {
    super(props)
    this.state = {
      icon: 'fa fa-chevron-down'
    }
  }
  iconType() {
    const { icon } = this.state
    if (icon === 'fa fa-chevron-down') {
      this.setState({ icon: 'fa fa-chevron-up' })
    } else {
      this.setState({ icon: 'fa fa-chevron-down' })
    }
  }

  render() {
    const { editor } = this.props
    const { icon } = this.state

    const data = removeEmpty(cloneDeep(editor.fullData))

    return (
      <Accordion defaultActiveKey="0" className="json-accordion">
        <Card>
          <Accordion.Toggle className="json-accordion-header" data-testid="json-view__accordion-header" as={Card.Header} eventKey="0" onClick={() => { this.iconType() }}>
            JSON
            <i className={icon} style={{ marginLeft: '620px' }} />
          </Accordion.Toggle>
          <Accordion.Collapse eventKey="0">
            <Card.Body id="json-text-field" data-testid="json-view__full-data">
              <JSONPretty data={data} />
            </Card.Body>
          </Accordion.Collapse>
        </Card>
      </Accordion>
    )
  }
}
export default observer(JSONView)
