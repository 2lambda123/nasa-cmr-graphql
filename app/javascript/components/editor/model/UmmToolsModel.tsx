/* eslint-disable dot-notation */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { UiSchema } from '@rjsf/utils'
import toolInformationUiSchema from '../data/configuration/uiSchemas/tools/tool_information'
import relatedUrlsUiSchema from '../data/configuration/uiSchemas/tools/related_urls'
import toolsConfiguration from '../data/configuration/uiForms/toolsConfiguration'
import organizationUiSchema from '../data/configuration/uiSchemas/tools/organization'
import UmmModel from './UmmModel'
import descriptiveKeywordsUiSchema from '../data/configuration/uiSchemas/tools/descriptive_keywords'
import potentialActionUiSchema from '../data/configuration/uiSchemas/tools/potential_action'
import toolContactsUISchema from '../data/configuration/uiSchemas/tools/tool_contacts'
import compatibilityAndUsabilityUiSchema from '../data/configuration/uiSchemas/tools/compatibility_and_usability'
import CustomArrayFieldTemplate from '../components/templates/CustomArrayFieldTemplate'
import schema from '../data/schemas/umm_tools_schema'

export default class UmmToolsModel extends UmmModel {
  constructor() {
    super(schema)
  }
  get documentTypeForDisplay() {
    return 'Tool Record'
  }
  get documentType() {
    return 'tool_drafts'
  }
  get formSections() {
    return toolsConfiguration
  }
  getFormSchema(): { [key: string]: object } {
    // use this function to edit schema
    return super.getFormSchema()
  }
  get uiSchema() {
    const base: UiSchema = {
      'ui:submitButtonOptions': { norender: true, submitText: 'Save' },
      'ui:ArrayFieldTemplate': CustomArrayFieldTemplate
    }
    if (this.currentSection.displayName === 'Related URLs') {
      const uiSchema: any = relatedUrlsUiSchema
      const urlUiSchema: any = uiSchema.RelatedURLs.items
      urlUiSchema['ui:service'] = this.service

      // Format
      const formatUiSchema: any = uiSchema.RelatedURLs.items.Format
      formatUiSchema['ui:service'] = this.service
      formatUiSchema['ui:keyword_scheme'] = 'granule_data_format'
      formatUiSchema['ui:keyword_scheme_column_names'] = ['short_name']

      // Mime Type
      const mimeTypeUiSchema: any = uiSchema.RelatedURLs.items.MimeType
      mimeTypeUiSchema['ui:service'] = this.service
      mimeTypeUiSchema['ui:keyword_scheme'] = 'mime_type'
      mimeTypeUiSchema['ui:keyword_scheme_column_names'] = ['mime_type']
      return { ...uiSchema, ...base }
    }

    if (this.currentSection.displayName === 'Tool Information') {
      const uiSchema: any = toolInformationUiSchema
      const urlUiSchema: any = uiSchema.URL
      urlUiSchema['ui:service'] = this.service
      urlUiSchema['ui:keyword_scheme'] = 'rucontenttype'
      return { ...uiSchema, ...base }
    }
    if (this.currentSection.displayName === 'Descriptive Keywords') {
      const uiSchema = descriptiveKeywordsUiSchema
      const toolKeywords: any = uiSchema.ToolKeywords
      toolKeywords['ui:service'] = this.service

      toolKeywords['ui:keyword_scheme'] = 'science_keywords'
      toolKeywords['ui:picker_title'] = 'TOOL KEYWORD'
      toolKeywords['ui:scheme_values'] = [
        'ToolCategory', 'ToolTopic', 'ToolTerm', 'ToolSpecificTerm'
      ]
      toolKeywords['ui:keyword_scheme_column_names'] = ['toolkeywords', 'category', 'topic', 'term', 'variable_level_1', 'variable_level_2', 'variable_level_3']
      toolKeywords['ui:filter'] = (path: string[]) => path[1] !== 'EARTH SCIENCE'

      return { ...uiSchema, ...base }
    }

    if (this.currentSection.displayName === 'Tool Organizations') {
      const uiSchema: any = organizationUiSchema
      const organizations: any = uiSchema.Organizations.items
      organizations['ui:service'] = this.service
      organizations['ui:keyword_scheme'] = 'providers'
      organizations['ui:keyword_scheme_column_names'] = ['short_name', 'long_name', 'url']
      return { ...uiSchema, ...base }
    }

    if (this.currentSection.displayName === 'Tool Contacts') {
      const uiSchema: any = toolContactsUISchema
      return { ...uiSchema, ...base }
    }

    if (this.currentSection.displayName === 'Potential Action') {
      const uiSchema: any = potentialActionUiSchema
      return { ...uiSchema, ...base }
    }
    if (this.currentSection.displayName === 'Compatibility And Usability') {
      const uiSchema: any = compatibilityAndUsabilityUiSchema
      return { ...uiSchema, ...base }
    }
    return { ...base }
  }

  // In this case the old preview sends the same section name that the react component uses (just lowercase), so
  // nothing needs to be migrated. once the old MMT preview is replaced, this 'migrate" functionality can be removed.
  migratedSectionName(sectionName: string): string {
    return sectionName
  }
}
