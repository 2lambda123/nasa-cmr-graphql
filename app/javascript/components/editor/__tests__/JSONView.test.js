import {
  fireEvent, render, screen
} from '@testing-library/react'
import React from 'react'
import { BrowserRouter } from 'react-router-dom'
import JSONView from '../components/JSONView'
import MetadataEditor from '../MetadataEditor'
import UmmToolsModel from '../model/UmmToolsModel'

describe('JSON View', () => {
  it('renders the JSON View with correct data', async () => {
    const model = new UmmToolsModel()
    model.fullData = {
      Name: 'JSON VIEW TEST',
      LongName: 'long name',
      Version: '1'
    }
    const editor = new MetadataEditor(model)
    const props = {
      editor
    }
    const { container, getByTestId } = render(
      <BrowserRouter>
        <JSONView {...props} />
      </BrowserRouter>
    )
    expect(getByTestId('json-view__full-data')).toHaveTextContent('{ "Name": "JSON VIEW TEST", "LongName": "long name", "Version": "1" }')
    expect(container).toMatchSnapshot()
  })

  it('testing Accordion Toggle', async () => {
    const model = new UmmToolsModel()
    model.fullData = {
      Name: 'JSON VIEW TEST',
      LongName: 'long name',
      Version: '1'
    }
    const editor = new MetadataEditor(model)
    const props = {
      editor
    }
    const { container } = render(
      <BrowserRouter>
        <JSONView {...props} />
      </BrowserRouter>
    )
    const accordionBtn = screen.getByTestId('json-view__accordion-header')

    // closes the accordion
    fireEvent.click(await accordionBtn)

    // opens the accordion
    fireEvent.click(await accordionBtn)
    expect(container).toMatchSnapshot()
  })
})
