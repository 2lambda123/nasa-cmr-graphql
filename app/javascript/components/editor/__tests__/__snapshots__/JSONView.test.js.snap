// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSON View renders the JSON View with correct data 1`] = `
<div>
  <div
    class="json-accordion accordion"
  >
    <div
      class="card"
    >
      <div
        class="json-accordion-header card-header"
        data-testid="json-view__accordion-header"
      >
        JSON
        <i
          class="fa fa-chevron-down"
          style="margin-left: 620px;"
        />
      </div>
      <div
        class="collapse show"
      >
        <div
          class="card-body"
          data-testid="json-view__full-data"
          id="json-text-field"
        >
          <div>
            <pre
              class="__json-pretty__"
            >
              {
  "
              <span
                class="__json-key__"
              >
                Name
              </span>
              ": 
              <span
                class="__json-string__"
              >
                "JSON VIEW TEST"
              </span>
              ,
  "
              <span
                class="__json-key__"
              >
                LongName
              </span>
              ": 
              <span
                class="__json-string__"
              >
                "long name"
              </span>
              ,
  "
              <span
                class="__json-key__"
              >
                Version
              </span>
              ": 
              <span
                class="__json-string__"
              >
                "1"
              </span>
              
}
            </pre>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`JSON View testing Accordion Toggle 1`] = `
<div>
  <div
    class="json-accordion accordion"
  >
    <div
      class="card"
    >
      <div
        class="json-accordion-header card-header"
        data-testid="json-view__accordion-header"
      >
        JSON
        <i
          class="fa fa-chevron-down"
          style="margin-left: 620px;"
        />
      </div>
      <div
        class="collapsing"
        style="height: 0px;"
      >
        <div
          class="card-body"
          data-testid="json-view__full-data"
          id="json-text-field"
        >
          <div>
            <pre
              class="__json-pretty__"
            >
              {
  "
              <span
                class="__json-key__"
              >
                Name
              </span>
              ": 
              <span
                class="__json-string__"
              >
                "JSON VIEW TEST"
              </span>
              ,
  "
              <span
                class="__json-key__"
              >
                LongName
              </span>
              ": 
              <span
                class="__json-string__"
              >
                "long name"
              </span>
              ,
  "
              <span
                class="__json-key__"
              >
                Version
              </span>
              ": 
              <span
                class="__json-string__"
              >
                "1"
              </span>
              
}
            </pre>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
